Hash table

tebeka opened the pull request:
Initial code for a simple hash table implementation in Go

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+	numBuckets = 256 // Number of buckets in hash
+)
+
+// entry in the hash table
```
All comments need to be proper sentence structure and complete sentences. Since this is teaching material, we will comment more than not. Make sure the comments are being very clear.

// Entry represents a value we will store in the hash table.

tebeka replied:
OK. IMO the upper case `Entry` might confuse.

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+// Set sets the key to value in the hash
+func (h *Hash) Set(key string, value int) {
+	b := h.hashKey(key)
+	// Look for existing key
```
Always a line space above a comment.

tebeka replied:
OK. Why?

ardan-bkennedy replied:
For readability and that is the style we use in this code and all the code I write and teach.

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+// Hash is a simple Hash table
+type Hash struct {
+	buckets [][]*entry
+	h       maphash.Hash
```
This variable `h` doesn't work for readability.  Later on you have code that looks like `h.h.Reset()`

I was hope we could use embedding but that won't work. :(

I think you call it `hash`. then you will have `h.hash.Reset()`

tebeka replied:
OK

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+
+// Set sets the key to value in the hash
+func (h *Hash) Set(key string, value int) {
+	b := h.hashKey(key)
```
It seems this hashkey guarentees a value inside the length of the slice?

tebeka replied:
Yup

ardan-bkennedy replied:
Can we explain how that works?

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// hashKey returns the bucket index for key
+func (h *Hash) hashKey(key string) int {
```
You need to explain how this hash function maintains a value from 0 to len(h.buckets).

tebeka replied:
OK

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+// NewHash returns a new hash table
+func NewHash() *Hash {
+	return &Hash{
+		buckets: make([][]*entry, numBuckets),
```
Why do we need to use pointer semantics in the collection. I am going back and forth on this. If the hash table is holding the original I would want to use value but if this is a secondary index, then pointer would be good.  Not Sure.

tebeka replied:
The pointer semantics is for `Set`. It's either what's currently in the code have or:
```go
	b := h.hashKey(key)

	// Look for existing key
	for i := range h.buckets[b] {
		if h.buckets[i].key == key {
			(&h.buckets[i]).value = value
			return
		}
	}

	h.buckets[b] = append(h.buckets[b], &entry{key, value})
```

The line `(&h.buckets[i]).value = value` is tricky. Using the following won't work.
```go
for _, e := range h.buckets[b] {
		if e.key == key {
			(&e).value = value
			return
```

Your call, I'm good with both.

ardan-bkennedy replied:
Can we use value semantic mutation and just replace the value?

ardan-bkennedy replied:
I feel like I am missing something. We could just copy over the value for the new one?

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+		}
+	}
+
+	h.buckets[b] = append(h.buckets[b], &entry{key, value})
```
Now I defintely want to use value semantics since we own the value. Keep the data closer together.

tebeka replied:
See above comment

ardan-bkennedy replied:
We can use value semantic mutation and just replace the value.

tebeka replied:
OK, changed with a TODO for a good comment on why we create a new entry

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+		if e.key == key {
+			copy(bucket[i:], bucket[i+1:])
+			bucket = bucket[:len(bucket)-1]
+			// Free memory when bucket shrinks a lot
```
Line space, proper sentence structure.

Also, this is teaching material, we need more comments walking the reader through the code. Assume someone looking at code like this for the first time. I know it will make the functions longer, but the purpose is to teach.

tebeka replied:
OK

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+func (h *Hash) Set(key string, value int) {
+	b := h.hashKey(key)
+	// Look for existing key
+	for _, e := range h.buckets[b] {
```
Wondering if we don't use `entry` over `e` for clarity. This is complicated stuff.

Actually. If there is a comment that says this the variables become self evident.

// Iterate over the entries for the specified bucket.

tebeka replied:
OK

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+	bucket := h.buckets[b]
+	for i, e := range bucket {
+		if e.key == key {
+			copy(bucket[i:], bucket[i+1:])
```
We need comments for all this stuff.

tebeka replied:
Created `removeEntry(bucket []*entry, i int) []*entry` function

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// Len return the number of elements in the hash
+func (h *Hash) Len() int {
```
This seems expensive. Do real implementations keep this meta-data?

ardan-bkennedy replied:
Maybe we don't want to teach keeping meta-data?

tebeka replied:
This is expensive, and real implementations probably keep length metadata. However it's a bit tricky to keep it updated. In `Delete` it obvious but in `Set` it's only if it's a new key. I prefer the simpler implementation.

ardan-bkennedy replied:
Ok, maybe we leave a comment in the function header?

tebeka replied:
Added a TODO comment.

ardan-bkennedy commented on topics/go/algorithms/data/hash/hash_test.go:
```diff
@@ -0,0 +1,63 @@
+	"testing"
+)
+
+func TestFunction(t *testing.T) {
```
The tests need to follow the semantics/mechanics found here for consistency.

https://github.com/ardanlabs/gotraining/blob/master/topics/go/algorithms/strings/reverse/reverse_test.go

tebeka replied:
OK

tebeka replied:
Is this the right link? Not sure how to do table driven tests here (in a simple manner).

ardan-bkennedy replied:
We don't need to do a table test, jsut the rest of the formating.

tebeka replied:
Let's fix this when coding together

ardan-bkennedy commented:
It might be better to pair on this code together using VS Code. What do you think?

tebeka commented:
> It might be better to pair on this code together using VS Code. What do you think?

Sure. I'm pretty open. LMK when.

tebeka commented:
@ardan-bkennedy When I've deleted the `student` folder I also deleted the following diagram that IMO will help. Should be in `hash.go` top documentaton:

```
Hash Data Diagram


       hashKey(key) ────────┐
			    │
			    ↓
	┌────┬─────┬─────┬────┬─────┬─────┬─────┬─────┐
	│    │     │     │    │     │     │     │     │  ←── buckets
	└────┴─────┴─────┴────┴─────┴─────┴─────┴─────┘
	   │                │
	   │                │
	   ↓                ↓
       ┌─────────────┐   ┌─────────────┐
       │ key │ value │   │ key │ value │  ←── entry
       ├─────────────┤   ├─────────────┤
       │ key │ value │   │ key │ value │
       ├─────────────┤   └─────────────┘
       │ key │ value │
       ├─────────────┤
       │ key │ value │
       ├─────────────┤
       │ key │ value │
       └─────────────┘
```

ardan-bkennedy commented:
Add a doc.go for all the documentation you want to add.

> On Apr 20, 2020, at 4:09 AM, Miki Tebeka <notifications@github.com> wrote:
> 
> 
> @ardan-bkennedy <https://github.com/ardan-bkennedy> When I've deleted the student folder I also deleted the following diagram that IMO will help. Should be in hash.go top documentaton:
> 
> Hash Data Diagram
> 
> 
>        hashKey(key) ────────┐
> 			    │
> 			    ↓
> 	┌────┬─────┬─────┬────┬─────┬─────┬─────┬─────┐
> 	│    │     │     │    │     │     │     │     │  ←── buckets
> 	└────┴─────┴─────┴────┴─────┴─────┴─────┴─────┘
> 	   │                │
> 	   │                │
> 	   ↓                ↓
>        ┌─────────────┐   ┌─────────────┐
>        │ key │ value │   │ key │ value │  ←── entry
>        ├─────────────┤   ├─────────────┤
>        │ key │ value │   │ key │ value │
>        ├─────────────┤   └─────────────┘
>        │ key │ value │
>        ├─────────────┤
>        │ key │ value │
>        ├─────────────┤
>        │ key │ value │
>        └─────────────┘
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/ardanlabs/gotraining/pull/315#issuecomment-616383846>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AARMUROEIFBIBVMUQBEGNTDRNP7MNANCNFSM4MKPQCIA>.
>

tebeka commented:
> Add a doc.go for all the documentation you want to add.

Done in #317
